lib/PipelineC/Tracing/Inspector.cpp

```diff
@@ -0,0 +1,98 @@
+/// \file Runner.cpp
+/// \brief Implements the logic for inspecting a trace file. Right now this
+///        entails listing and extracting buffers
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/Support/Base64.h"
+
+#include "revng/ADT/ConstexprString.h"
+#include "revng/PipelineC/PipelineC.h"
+#include "revng/PipelineC/Tracing/Common.h"
+#include "revng/PipelineC/Tracing/Trace.h"
+
+static llvm::StringMap<std::set<size_t>> BufferRegistry;
+
+template<ConstexprString Name, typename ArgTupleT, size_t I = 0>
+static void handleArguments() {
+  using ArgT = std::tuple_element_t<I, ArgTupleT>;
+  constexpr int LH = LengthHint<Name, I>;
+  if constexpr (anyOf<ArgT, char *, const char *>() && LH >= 0) {
+    std::string StrName(Name);
+    BufferRegistry[StrName].insert(I);
+  }
+
+  if constexpr (I + 1 < std::tuple_size_v<ArgTupleT>)
+    handleArguments<Name, ArgTupleT, I + 1>();
+}
+
+template<ConstexprString Name, typename ReturnT, typename... Args>
+static void handleFunction(std::function<ReturnT(Args...)> Function) {
+  if constexpr (sizeof...(Args) > 0)
+    handleArguments<Name, std::tuple<Args...>>();
+}
+
+template<ConstexprString Name, typename FunctionT>
+static void registerFunction(FunctionT Function) {
+  handleFunction<Name>(std::function{ Function });
+}
+
+static bool InitOk = []() {
+// Add autogenerated registerRunner calls to register the
+// other functions automatically
+#define FUNCTION(fname) registerFunction<#fname>(fname);
+#include "revng/PipelineC/Functions.inc"
+#undef FUNCTION
+  return true;
+}();
+
+inline llvm::Expected<std::vector<char>>
+extractBuffer(const revng::tracing::Argument &Arg) {
+  revng_assert(Arg.isScalar());
+  std::vector<char> Result;
+  llvm::Error Err = llvm::decodeBase64(Arg.getScalar(), Result);
+  if (Err)
+    return std::move(Err);
+
+  llvm::consumeError(std::move(Err));
+  return Result;
+}
```

Comment:
All of this code should become a class where `InitOk` is the constructor, `BufferRegistry` is a field and the others are methods.

Reply:
Done